src/atari800.man
.TH ATARI800 1 "2019-12-28" "Atari800 4.2.0"

.SH NAME
atari800 \- emulator of Atari 8-bit computers and the 5200 console

.SH SYNOPSIS
.B atari800
.RI [ option "]... [" file ]...

.SH DESCRIPTION
.B atari800
emulates the Atari 8-bit computer systems including the 400, 800, 1200XL,
600XL, 800XL, 65XE, 130XE, 800XE and the XE Game System, and also the Atari
5200 SuperSystem console.

.SH OPTIONS

.TP
.B \-help
Print complete and up-to-date list of command line switches
.TP
.B \-v
Print emulator version
.TP
.B \-verbose
Display framerate when exiting
.TP
.BI \-config\  filename
Specify an alternative configuration filename
.TP
.B \-autosave\-config
Automatically save the current configuration on emulator exit.
.TP
.B \-no\-autosave\-config
Don't save the current configuration on emulator exit (the default).

.TP
.BI \-osa_rom\  filename
Path to file containing Atari Rev.A Operating System.
.TP
.BI \-osb_rom\  filename
Path to file containing Atari Rev.B Operating System.
.TP
.BI \-xlxe_rom\  filename
Path to file containing Atari XL/XE Operating System.
.TP
.BI \-5200_rom\  filename
Path to file containing Atari 5200 Games System ROM.
.TP
.BI \-basic_rom\  filename
Path to file containing Atari BASIC ROM.
Used to override paths defined at compile time.

.TP
\fB\-800\-rev auto\fR|\fBa\-ntsc\fR|\fBa\-pal\fR|\fBb\-ntsc\fR|\fBcustom\fR|\fBaltirra\fR
Select operating system revision for Atari 800:
.PP
.RS
.PD 0
.TP
.B auto
The default.
Use the most appropriate from the available ones.
.TP
.B a\-ntsc
Use OS rev. A NTSC from early NTSC 400/800 units.
.TP
.B a\-pal
Use OS rev. A PAL from PAL 400/800 units.
.TP
.B b\-ntsc
Use OS rev. B NTSC from late NTSC 400/800 units.
.TP
.B custom
Use a custom OS.
.TP
.B altirra
Use AltirraOS that is included in the emulator.
.PD
.RE
.TP
\fB\-xl\-rev auto\fR|\fB10\fR|\fB11\fR|\fB1\fR|\fB2\fR|\fB3a\fR|\fB3b\fR|\fB5\fR|\fB3\fR|\fB4\fR|\fB59\fR|\fB59a\fR|\fBcustom\fR|\fBaltirra\fR
Select operating system revision for Atari XL/XE:
.PP
.RS
.PD 0
.TP
.B auto
The default.
Use the most appropriate from the available ones.
.TP
.B 10
Use OS AA000000 rev. 10 from early 1200XL units.
.TP
.B 11
Use OS AA000001 rev. 11 from late 1200XL units.
.TP
.B 1
Use OS BB000000 rev. 1 from the 600XL.
.TP
.B 2
Use OS BB000001 rev. 2 from the 800XL and early 65XE/130XE units.
.TP
.B 3a
Use prototype OS BB000002 rev. 3 from 1400XL/1450XLD units (also known as 1540OS3.V0 and 1450R3V0.ROM).
.TP
.B 3b
Use prototype OS BB000002 rev. 3 ver. 4 from 1400XL/1450XLD units (also known as os1450.128 and 1450R3VX.ROM).
.TP
.B 5
Use prototype OS CC000001 rev. 4 (also known as Rev. 5), for which sources are
available at <http://www.atariage.com/forums/topic/78579-a800ossrc/page__view__findpost__p__961535>.
.TP
.B 3
Use OS BB000001 rev. 3 from late 65XE/130XE units.
.TP
.B 4
Use OS BB000001 rev. 4 from the XEGS.
.TP
.B 59
Use OS BB000001 rev. 59 from the Arabic 65XE.
.TP
.B 59a
Use OS BB000001 rev. 59 from Kevin Savetz' Arabic 65XE: <http://www.savetz.com/vintagecomputers/arabic65xe/>.
.TP
.B custom
Use a custom OS.
.TP
.B altirra
Use AltirraOS that is included in the emulator.
.PD
.RE
.TP
\fB\-5200\-rev auto\fR|\fBorig\fR|\fBa\fR|\fBcustom\fR|\fBaltirra\fR
Select BIOS revision for Atari 5200:
.PP
.RS
.PD 0
.TP
.B auto
The default.
Use the most appropriate from the available ones.
.TP
.B orig
Use BIOS from 4-port and early 2\-port 5200 units.
.TP
.B a
Use BIOS rev. A from late 2\-port 5200 units.
.TP
.B custom
Use a custom BIOS.
.TP
.B altirra
Use Altirra BIOS that is included in the emulator.
.PD
.RE
.TP
\fB\-basic\-rev auto\fR|\fBa\fR|\fBb\fR|\fBc\fR|\fBcustom\fR|\fBaltirra\fR
Select BASIC revision:
.PP
.RS
.PD 0
.TP
.B auto
The default.
Use the most appropriate from the available ones.
.TP
.B a
Use BASIC rev. A.
.TP
.B b
Use BASIC rev. B from early 600XL/800XL units.
.TP
.B c
Use BASIC rev. C from late 600XL/800XL and all XE units.
.TP
.B custom
Use a custom BASIC.
.TP
.B altirra
Use Altirra BASIC that is included in the emulator.
.PD
.RE
.TP
\fB\-xegame\-rev auto\fR|\fBorig\fR|\fBcustom\fR
Select XEGS builtin game version:
.PP
.RS
.PD 0
.TP
.B auto
The default.
Use the most appropriate from the available ones.
.TP
.B orig
Use the original Missile Command game.
.TP
.B custom
Use a custom game ROM.
.PD
.RE

.TP
.B \-atari
Emulate Atari 400/800
.TP
.B \-1200
Emulate Atari 1200XL
.TP
.B \-xl
Emulate Atari 800XL
.TP
.B \-xe
Emulate Atari 130XE
.TP
.B \-320xe
Emulate Atari 320XE (Compy-Shop)
.TP
.B \-rambo
Emulate Atari 320XE (Rambo)
.TP
.B \-576xe
Emulate Atari 576XE
.TP
.B \-1088xe
Emulate Atari 1088XE
.TP
.B \-xegs
Emulate Atari XEGS
.TP
.B \-5200
Emulate Atari 5200
.TP
.B \-c
Enable RAM between 0xc000 and 0xcfff in Atari 800

.TP
.BI \-axlon\  n
Use Atari 800 Axlon memory expansion: \fIn\fR k total RAM
.TP
.B \-axlon0f
Use Axlon shadow at 0x0fc0-0x0fff
.TP
.BI \-mosaic\  n
Use 400/800 Mosaic memory expansion: \fIn\fR k total RAM
.TP
.B \-mapram
Enable the MapRAM memory expansion. Works only when emulating an Atari XL/XE machine.
.TP
.B \-no-mapram
Disable the MapRAM memory expansion.

.TP
.B \-pal
Emulate PAL TV mode
.TP
.B \-ntsc
Emulate NTSC TV mode

.TP
.B \-nobasic
Used to disable Basic when starting the emulator in XL/XE mode.
Simulates the Option key being held down during system boot.
.TP
.B \-basic
Turn on Atari BASIC ROM

.TP
.BI \-cart\  filename
Insert cartridge (CART or raw format)
.TP
.BR "\-cart\-type 0" .. 75
Select type of the cartridge inserted with the \fB\-cart\fR option.
When inserting a raw ROM image its type cannot be detected automaticcaly,
and must be provided with this option.
Setting to \fB0\fR means disabling the cartridge, and any other number
indicates a specific cartridge type:
.RS
.PD 0
.TP
.B 1
Standard 8 KB cartridge
.TP
.B 2
Standard 16 KB cartridge
.TP
.B 3
OSS two chip 16 KB cartridge (034M)
.TP
.B 4
Standard 32 KB 5200 cartridge
.TP
.B 5
DB 32 KB cartridge
.TP
.B 6
Two chip 16 KB 5200 cartridge
.TP
.B 7
Bounty Bob Strikes Back 40 KB 5200 cartridge
.TP
.B 8
64 KB Williams cartridge
.TP
.B 9
Express 64 KB cartridge
.TP
.B 10
Diamond 64 KB cartridge
.TP
.B 11
SpartaDos X 64 KB cartridge
.TP
.B 12
XEGS 32 KB cartridge
.TP
.B 13
XEGS 64 KB cartridge (banks 0-7)
.TP
.B 14
XEGS 128 KB cartridge
.TP
.B 15
OSS one chip 16 KB cartridge
.TP
.B 16
One chip 16 KB 5200 cartridge
.TP
.B 17
Decoded Atrax 128 KB cartridge
.TP
.B 18
Bounty Bob Strikes Back 40 KB cartridge
.TP
.B 19
Standard 8 KB 5200 cartridge
.TP
.B 20
Standard 4 KB 5200 cartridge
.TP
.B 21
Right slot 8 KB cartridge
.TP
.B 22
32 KB Williams cartridge
.TP
.B 23
XEGS 256 KB cartridge
.TP
.B 24
XEGS 512 KB cartridge
.TP
.B 25
XEGS 1 MB cartridge
.TP
.B 26
MegaCart 16 KB cartridge
.TP
.B 27
MegaCart 32 KB cartridge
.TP
.B 28
MegaCart 64 KB cartridge
.TP
.B 29
MegaCart 128 KB cartridge
.TP
.B 30
MegaCart 256 KB cartridge
.TP
.B 31
MegaCart 512 KB cartridge
.TP
.B 32
MegaCart 1 MB cartridge
.TP
.B 33
Switchable XEGS 32 KB cartridge
.TP
.B 34
Switchable XEGS 64 KB cartridge
.TP
.B 35
Switchable XEGS 128 KB cartridge
.TP
.B 36
Switchable XEGS 256 KB cartridge
.TP
.B 37
Switchable XEGS 512 KB cartridge
.TP
.B 38
Switchable XEGS 1 MB cartridge
.TP
.B 39
Phoenix 8 KB cartridge
.TP
.B 40
Blizzard 16 KB cartridge
.TP
.B 41
Atarimax 128 KB Flash cartridge
.TP
.B 42
Atarimax 1 MB Flash cartridge (old)
.TP
.B 43
SpartaDos X 128 KB cartridge
.TP
.B 44
OSS 8 KB cartridge
.TP
.B 45
OSS two chip 16 KB cartridge (043M)
.TP
.B 46
Blizzard 4 KB cartridge
.TP
.B 47
AST 32 KB cartridge
.TP
.B 48
Atrax SDX 64 KB cartridge
.TP
.B 49
Atrax SDX 128 KB cartridge
.TP
.B 50
Turbosoft 64 KB cartridge
.TP
.B 51
Turbosoft 128 KB cartridge
.TP
.B 52
Ultracart 32 KB cartridge
.TP
.B 53
Low bank 8 KB cartridge
.TP
.B 54
SIC! 128 KB cartridge
.TP
.B 55
SIC! 256 KB cartridge
.TP
.B 56
SIC! 512 KB cartridge
.TP
.B 57
Standard 2 KB cartridge
.TP
.B 58
Standard 4 KB cartridge
.TP
.B 59
Right slot 4 KB cartridge
.TP
.B 60
Blizzard 32 KB cartridge
.TP
.B 61
MegaMax 2 MB cartridge
.TP
.B 62
The!Cart 128 MB cartridge
.TP
.B 63
Flash MegaCart 4 MB cartridge
.TP
.B 64
MegaCart 2 MB cartridge
.TP
.B 65
The!Cart 32 MB cartridge
.TP
.B 66
The!Cart 64 MB cartridge
.TP
.B 67
XEGS 64 KB cartridge (banks 8-15)
.TP
.B 68
Atrax 128 KB cartridge
.TP
.B 69
aDawliah 32 KB cartridge
.TP
.B 70
aDawliah 64 KB cartridge
.TP
.B 71
Super Cart 64 KB 5200 cartridge (32K banks)
.TP
.B 72
Super Cart 128 KB 5200 cartridge (32K banks)
.TP
.B 73
Super Cart 256 KB 5200 cartridge (32K banks)
.TP
.B 74
Super Cart 512 KB 5200 cartridge (32K banks)
.TP
.B 75
Atarimax 1 MB Flash cartridge (new)
.PD
.RE
If this option is not given, the user will be asked to choose the cartridge
type when the emulator starts.
.TP
.BI \-cart2\  filename
Insert piggyback cartridge (CART or raw format).
This works only if the first cartridge is a pass-through (currently only
SpartaDOS X 64KB and 128KB types).
.TP
.BR "\-cart2\-type 0" .. 75
Select type of the cartridge inserted with the \fB-cart2\fR option.
When inserting a raw ROM image its type cannot be detected automatically,
and must be provided with this option.
The available values are the same as for the \fB\-cart\-type\fR option above.
.TP
.B \-cart\-autoreboot
Automatically reboot after cartridge inserting/removing (this is the default
setting).
This does not apply to the piggyback cartridge - inserting or removing it
never causes automatic reboot.
.TP
.B \-no\-cart\-autoreboot
Disable automatic reboot after cartridge inserting/removing.

.TP
.BI \-run\  filename
Run Atari program (EXE, COM, XEX, BAS, LST)
.TP
.BI \-state\  filename
Load saved-state file
.TP
.BI \-tape\  filename
Attach cassette image (CAS format or raw file)
.TP
.BI \-boottape\  filename
Attach cassette image and boot it
.TP
.B \-tape\-readonly
Set the attached cassette image as read-only. 


.TP
.B \-1400
Emulate the Atari 1400XL
.TP
.B \-xld
Emulate the Atari 1450XLD
.TP
.B \-bb
Emulate the CSS Black Box
.TP
.B \-mio
Emulate the ICD MIO board

.TP
.B \-nopatch
Normally the OS is patched giving very fast I/O.
This options prevents the patch from being applied so that the OS accesses
the serial port hardware directly.
This option will probably never be needed since programs that access the
serial hardware should work even if the OS has been patched.
.TP
.B \-nopatchall
Don't patch OS at all, H:, P: and R: devices won't work

.TP
.BI \-H1\  path
Set path for H1: device
.TP
.BI \-H2\  path
Set path for H2: device
.TP
.BI \-H3\  path
Set path for H3: device
.TP
.BI \-H4\  path
Set path for H4: device
.TP
.BI \-Hpath\  path
Set path for Atari executables on the H: device
.TP
.B \-hreadonly
Enable read-only mode for H: device
.TP
.B \-hreadwrite
Disable read-only mode for H: device
.TP
.B \-devbug
Put debugging messages for H: and P: devices in log file

.TP
.B \-rtime
Enable R-Time 8 emulation
.TP
.B \-nortime
Disable R-Time 8 emulation

.TP
\fB\-rdevice\fR [\fIdev\fR]
Enable R: device.
If \fIdev\fR is specified then it's used as host serial device name (e.g.
\fI/dev/ttyS0\fR on linux).
If there is no \fIdev\fR specified then R: is directed to network.

.TP
.B \-mouse off
Do not use mouse
.TP
.B \-mouse pad
Emulate paddles
.TP
.B \-mouse touch
Emulate Atari Touch Tablet
.TP
.B \-mouse koala
Emulate Koala Pad
.TP
.B \-mouse pen
Emulate Light Pen
.TP
.B \-mouse gun
Emulate Light Gun
.TP
.B \-mouse amiga
Emulate Amiga mouse
.TP
.B \-mouse st
Emulate Atari ST mouse
.TP
.B \-mouse trak
Emulate Atari Trak-Ball
.TP
.B \-mouse joy
Emulate joystick using mouse
.TP
.BI \-mouseport\  num
Set mouse port 1-4 (default 1)
.TP
.BI \-mousespeed\  num
Set mouse speed 1-9 (default 3)
.TP
.B \-multijoy
Emulate MultiJoy4 interface
.TP
.B \-directmouse
Use mouse's absolute position
.TP
.BI \-cx85\  num
Emulate CX85 numeric keypad on port \fInum\fR
.TP
.B \-grabmouse
SDL only, prevent mouse pointer from leaving the window

.TP
.BI \-record\  filename
Record all input events to \fIfilename\fR. Can be used for gaming contests
(highest score etc).
.TP
.BI \-playback\  filename
Playback input events from \fIfilename\fR. Watch an expert play the game.

.TP
.B \-refresh
Controls screen refresh rate.
A numerical value follows this option which specifies how many emulated
screen updates are required before the actual screen is updated.
This value effects the speed of the emulation: A higher value results in
faster CPU emulation but a less frequently updated screen.

.TP
\fB\-ntsc\-artif \fImode\fR, \fB\-pal\-artif \fImode\fR
Set emulation mode of video artifacts in NTSC or PAL, respectively. The
available values for \fImode\fR are:
.RS
.PD 0
.TP
.B none
Disable video artifacts.
.TP
.B ntsc-old
Simple emulation of NTSC composite video artifacts. Fast but inaccurate.
.TP
.B ntsc-new
Improved emulation of NTSC artifacts. May look better than \fBntsc\-old\fR.
.TP
.B ntsc-full
Full emulation of NTSC artifacts. Rather slow and available only in 16-
and 32-bit video modes.
.TP
.B pal-simple
Simple emulation of PAL chroma blending, without composite artifacts. Fast but
inaccurate.
.TP
.B pal-blend
Accurate emulation of PAL chroma blending, without composite artifacts.
Available only in 16- and 32-bit video modes.
.PD
.RE
.TP
.BI \-artif\  mode
Set artifacting mode 0-4 (0 = disable). Only for tv effects \fBntsc\-old\fR and \fBntsc\-new\fR.

.TP
.BR "\-colors\-preset standard" | "deep\-black" | vibrant
Use one of predefined colour adjustments
.TP
.BI \-saturation\  n
Set screen color saturation (like TV Colour control)
.TP
\fB\-ntsc\-saturation \fIn\fR, \fB\-pal\-saturation \fIn\fR
Set saturation only for NTSC or PAL, respectively
.TP
.BI \-contrast\  n
Set screen contrast (also called white level)
.TP
\fB\-ntsc\-contrast \fIn\fR, \fB\-pal\-contrast \-fIn\fR
Set contrast only for NTSC or PAL, respectively
.TP
.BI \-brightness\  n
Set screen brightness (also called black level)
.TP
\fB\-ntsc\-brightness \fIn\fR, \fB\-pal\-brightness \fIn\fR
Set brightness only for NTSC or PAL, respectively
.TP
.BI \-gamma\  n
Set screen gamma correction
.TP
\fB\-ntsc\-gamma \fIn\fR, \fB\-pal\-gamma \fIn\fR
Set gamma adjustment only for NTSC or PAL, respectively
.TP
.BI \-tint\  n
Set tint -1..1.
.TP
\fB\-ntsc\-tint \fIn\fR, \fB\-pal\-tint \fIn\fR
Set tint only for NTSC or PAL, respectively
.TP
.BI \-ntsc-colordelay\  n
Set GTIA color delay for NTSC system.
This emulates adjusting the potentiometer existing at the bottom of Atari
computers, which adjusts hues of colors produced by the computer.
.TP
.BI \-pal\-colordelay\  n
Set GTIA color delay for PAL system.

.TP
\fB\-paletten \fIfilename\fR, \fB\-palettep \fIfilename\fR
Read Atari NTSC/PAL colors from ACT file
.TP
.BR \-paletten\-adjust ,\  \-palettep\-adjust
Apply colour adjustments (brightness, contrast etc.) to the loaded NTSC/PAL
palette (by default the loaded palette is displayed unmodified).

.TP
.BI \-screenshots\  pattern
Set filename pattern for screenshots.
Use to override the default pattern of \fIatari###.png\fR which produces
\fIatari000.png\fR, \fIatari001.png\fR etc. filenames.
Hashes are replaced with raising numbers.
Existing files are overwritten only if all the files defined by the pattern
exist.

.TP
.B \-showspeed
Show percentage of actual speed

.TP
.B \-sound
Enable sound
.TP
.B \-nosound
Disable sound
.TP
.BI \-dsprate\  freq
Set sound output frequency in Hz.
The default is 44100 Hz.
.TP
.B \-stereo
Enable stereo sound
.TP
.B \-nostereo
Disable stereo sound
.TP
.B \-audio16
Set sound output format to 16-bit
.TP
.B \-audio8
Set sound output format to 8-bit
.TP
.BI \-aname\  pattern
Set filename pattern for audio recordings.
Use to override the default pattern of \fIatari###.wav\fR which produces
\fIatari000.wav\fR, \fIatari001.wav\fR etc. filenames.
Hashes are replaced with raising numbers.

Note that WAV format files can support all audio codecs, including MP3,
but many programs assume WAV files contain only PCM audio.

If MP3 support was enabled when compiling the emulator, and MP3 audio is
selected using the \fB\-acodec mp3\fR option below, the default pattern
will be \fIatari###.mp3\fR to save in MP3 format files.
.TP
\fB\-acodec auto\fR|\fBpcm\fR|\fBmp3\fR|\fBmulaw\fR|\fBpcm_mulaw\fR|\fBadpcm\fR|\fBadpcm_ima_wav\fR|\fBadpcm_yahama\fR|\fBadpcm_ms\fR
Select the audio codec used when saving to AVI or WAV files. Some codecs are
lossy, meaning they reduce storage space while attempting to sound as close as
possible to the original audio.

.PP
.RS
.PD 0
.TP
.B auto
The default.
Use the codec that provides the best audio quality, which is PCM.
.TP
.B pcm

python/pyatari800/atari800.py
import ctypes
import time

import numpy as np

from . import libatari800 as liba8
from . import generic_interface as g
from . import akey
from .save_state_parser import parse_state
from .colors import NTSC
from .emulator_base import EmulatorBase

import logging
logging.basicConfig(level=logging.WARNING)
log = logging.getLogger(__name__)


debug_frames = False

def debug_video(mem):
    offset = 336*24 + 128
    for y in range(32):
        print "%x:" % offset,
        for x in range(8,60):
            c = mem[x + offset]
            if (c == 0 or c == '\x00'):
                print " ",
            elif (c == 0x94 or c == '\x94'):
                print ".",
            elif (c == 0x9a or c == '\x9a'):
                print "X",
            else:
                try:
                    print ord(c),
                except TypeError:
                    print repr(c),
        print
        offset += 336;

def clamp(val):
    if val < 0.0:
        return 0
    elif val > 255.0:
        return 255
    return int(val)

ntsc_iq_lookup = [
    [  0.000,  0.000 ],
    [  0.144, -0.189 ],
    [  0.231, -0.081 ],
    [  0.243,  0.032 ],
    [  0.217,  0.121 ],
    [  0.117,  0.216 ],
    [  0.021,  0.233 ],
    [ -0.066,  0.196 ],
    [ -0.139,  0.134 ],
    [ -0.182,  0.062 ],
    [ -0.175, -0.022 ],
    [ -0.136, -0.100 ],
    [ -0.069, -0.150 ],
    [  0.005, -0.159 ],
    [  0.071, -0.125 ],
    [  0.124, -0.089 ],
    ]

def gtia_ntsc_to_rgb_table(val):
    # This is a better representation of the NTSC colors using a lookup table
    # rather than the phase calculations. Also from the same thread:
    # http://atariage.com/forums/topic/107853-need-the-256-colors/page-2#entry1319398
    cr = (val >> 4) & 15;
    lm = val & 15;

    y = 255*(lm+1)/16;
    i = ntsc_iq_lookup[cr][0] * 255
    q = ntsc_iq_lookup[cr][1] * 255

    r = y + 0.956*i + 0.621*q;
    g = y - 0.272*i - 0.647*q;
    b = y - 1.107*i + 1.704*q;

    return clamp(r), clamp(g), clamp(b)

def ntsc_color_map():
    rmap = np.empty(256, dtype=np.uint8)
    gmap = np.empty(256, dtype=np.uint8)
    bmap = np.empty(256, dtype=np.uint8)

    for i in range(256):
        r, g, b = gtia_ntsc_to_rgb_table(i)
        rmap[i] = r
        gmap[i] = g
        bmap[i] = b

    return rmap, gmap, bmap


def start_monitor_event_loop(emu):
    liba8.get_current_state(emu.output)
    a, p, sp, x, y, _, pc = emu.cpu_state
    print("In start_monitor_event_loop: A=%02x X=%02x Y=%02x SP=%02x FLAGS=%02x PC=%04x" % (a, x, y, sp, p, pc))

    # Confirm validity by checking raw locations
    emu.debug_state()
    liba8.monitor_step()
    #time.sleep(.5)


class Atari800(EmulatorBase):
    cpu = "6502"
    name = "atari800"
    pretty_name = "Atari 800"

    input_array_dtype = g.INPUT_DTYPE
    output_array_dtype = g.OUTPUT_DTYPE
    width = g.VIDEO_WIDTH
    height = g.VIDEO_HEIGHT

    def compute_color_map(self):
        self.rmap, self.gmap, self.bmap = ntsc_color_map()

    def calc_screens(self):
        rgb = np.empty((self.height, self.width, 3), np.uint8)
        rgba = np.empty((self.height, self.width, 4), np.uint8)
        return rgb, rgba

    @property
    def raw_array(self):
        return self.output.view(dtype=np.uint8)

    @property
    def video_array(self):
        return self.output['video'][0]

    @property
    def audio_array(self):
        return self.output['audio'][0]

    @property
    def state_array(self):
        return self.output['state'][0]

    @property
    def current_frame_number(self):
        return self.output['frame_number'][0]

    def begin_emulation(self, emu_args=None, event_loop=None, event_loop_args=None):
        self.args = self.normalize_args(emu_args)
        if event_loop is None:
            event_loop = start_monitor_event_loop
        if event_loop_args is None:
            event_loop_args = self
        liba8.start_emulator(self.args, event_loop, event_loop_args)
        liba8.prepare_arrays(self.input, self.output)
        self.parse_state()
        self.cpu_state = self.calc_cpu_data_array()

    def normalize_args(self, args):
        if args is None:
            args = [
                "-basic",
                #"-shmem-debug-video",
                #"jumpman.atr"
            ]
        return args

    def parse_state(self):
        base = np.byte_bounds(self.output)[0]
        self.video_start_offset = np.byte_bounds(self.video_array)[0] - base
        self.audio_start_offset = np.byte_bounds(self.audio_array)[0] - base
        self.state_start_offset = np.byte_bounds(self.state_array)[0] - base
        self.offsets, self.names, self.segments, self.segment_starts, self.computed_dtypes = parse_state(self.output['state'], self.state_start_offset)
        self.segments[0:0] = [
            (self.video_start_offset, self.video_start_offset + len(self.video_array), 0, "Video Frame"),
            (self.audio_start_offset, self.audio_start_offset + len(self.audio_array), 0, "Audio Data"),
        ]

    def end_emulation(self):
        pass

    def debug_video(self):
        debug_video(self.output[0]['video'].view(dtype=np.uint8))

    def debug_state(self):
        a, p, sp, x, y, _, pc = self.cpu_state

        # Confirm validity by checking raw locations
        names = self.names
        raw = self.raw_array
        assert a == raw[names['CPU_A']]
        assert p == raw[names['CPU_P']]
        assert sp == raw[names['CPU_S']]
        assert x == raw[names['CPU_X']]
        assert y == raw[names['CPU_Y']]
        assert pc == (raw[names['PC']] + 256 * raw[names['PC'] + 1])

    def next_frame(self):
        print("requesting next frame")
        self.process_key_state()
        self.frame_count += 1
        liba8.next_frame(self.input, self.output)
        print("frame finished=%s breakpoint=%s" % (self.output['frame_finished'][0], self.output['breakpoint_hit'][0]))
        self.process_frame_events()
        self.save_history()

    def process_frame_events(self):
        still_waiting = []
        for count, callback in self.frame_event:
            if self.frame_count >= count:
                print "processing %s", callback
                callback()
            else:
                still_waiting.append((count, callback))
        self.frame_event = still_waiting

    def calc_cpu_data_array(self):
        names = self.names
        computed_dtype = np.dtype([
            ("A", np.uint8),
            ("P", np.uint8),
            ("SP", np.uint8),
            ("X", np.uint8),
            ("Y", np.uint8),
            ("_", np.uint8, names['PC'] - names['CPU_A'] - 5),
            ("PC", '<u2'),
            ])
        raw = self.raw_array[names['CPU_A']:names['CPU_A'] + computed_dtype.itemsize]
        print("sizeof raw_array=%d raw=%d dtype=%d" % (len(self.raw_array), len(raw), computed_dtype.itemsize))
        dataview = raw.view(dtype=computed_dtype)
        return dataview[0]

    def calc_dtype_data(self, segment_name):
        d = np.dtype(self.computed_dtypes[segment_name])
        start = self.segment_starts[segment_name]
        raw = self.raw_array[start:start + d.itemsize]
        return raw.view(dtype=d)[0]

    # Utility functions

    def load_disk(self, drive_num, pathname):
        liba8.load_disk(drive_num, pathname)

    def save_history(self):
        # History is saved in a big list, which will waste space for empty
        # entries but makes things extremely easy to manage. Simply delete
        # a history entry by setting it to NONE.
        frame_number = self.output['frame_number'][0]
        if self.frame_count % 10 == 0:
            d = self.output.copy()
        else:
            d = None
        if len(self.history) != frame_number:
            print("frame %d: history out of sync. has=%d expecting=%d" % (frame_number, len(self.history), frame_number))
        self.history.append(d)
        if d is not None:
            self.print_history(frame_number)

    def restore_history(self, frame_number):
        print("restoring state from frame %d" % frame_number)
        if frame_number < 0:
            return
        d = self.history[frame_number]
        liba8.restore_state(d)
        self.history[frame_number + 1:] = []  # remove frames newer than this
        print("  %d items remain in history" % len(self.history))
        self.frame_event = []

    def print_history(self, frame_number):
        d = self.history[frame_number]
        print "history[%d] of %d: %d %s" % (d['frame_number'], len(self.history), len(d), d['state'][0][0:8])

    def get_previous_history(self, frame_cursor):
        n = frame_cursor - 1
        while n > 0:
            if self.history[n] is not None:
                return n
            n -= 1
        raise IndexError("No previous frame")

    def get_next_history(self, frame_cursor):
        n = frame_cursor + 1
        while n < len(self.history):
            if self.history[n] is not None:
                return n
            n += 1
        raise IndexError("No next frame")

    def get_color_indexed_screen(self, frame_number=-1):
        if frame_number < 0:
            output = self.output
        else:
            output = self.history[frame_number]
        raw = output['video'].reshape((self.height, self.width))
        #print "get_raw_screen", frame_number, raw
        return raw

    def get_frame_rgb(self, frame_number=-1):
        raw = self.get_color_indexed_screen(frame_number)
        self.screen_rgb[:,:,0] = self.rmap[raw]
        self.screen_rgb[:,:,1] = self.gmap[raw]
        self.screen_rgb[:,:,2] = self.bmap[raw]
        return self.screen_rgb

    def get_frame_rgba(self, frame_number=-1):
        raw = self.get_color_indexed_screen(frame_number)
        self.screen_rgba[:,:,0] = self.rmap[raw]
        self.screen_rgba[:,:,1] = self.gmap[raw]
        self.screen_rgba[:,:,2] = self.bmap[raw]
        self.screen_rgba[:,:,3] = 255
        return self.screen_rgba

    def get_frame_rgba_opengl(self, frame_number=-1):
        raw = np.flipud(self.get_color_indexed_screen(frame_number))
        self.screen_rgba[:,:,0] = self.rmap[raw]
        self.screen_rgba[:,:,1] = self.gmap[raw]
        self.screen_rgba[:,:,2] = self.bmap[raw]
        self.screen_rgba[:,:,3] = 255
        return self.screen_rgba

    ##### Input routines

    def send_char(self, key_char):
        self.input['keychar'] = key_char
        self.input['keycode'] = 0
        self.input['special'] = 0

    def send_keycode(self, keycode):
        self.input['keychar'] = 0
        self.input['keycode'] = keycode
        self.input['special'] = 0

    def send_special_key(self, key_id):
        self.input['keychar'] = 0
        self.input['keycode'] = 0
        self.input['special'] = key_id
        if key_id in [2, 3]:
            self.frame_event.append((self.frame_count + 2, self.clear_keys))

    def clear_keys(self):
        self.input['keychar'] = 0
        self.input['keycode'] = 0
        self.input['special'] = 0

    def set_option(self, state):
        self.input['option'] = state

    def set_select(self, state):
        self.input['select'] = state

    def set_start(self, state):
        self.input['start'] = state

    def process_key_state(self):
        pass

    ##### debugger convenience functions

    def enter_debugger(self):
        if self.active_event_loop is not None:
            print("Only one debugger may be active at a time!")
        else:
            print("Requesting debugger start via AKEY_UI")
            self.send_special_key(akey.AKEY_UI)

    def leave_debugger(self):
        liba8.monitor_clear()
        self.restart_cpu()

    def restart_cpu(self):
        if self.active_event_loop is not None:
            self.clear_keys()
            self.active_event_loop.Exit()
            print("alternate event loop is over.")
            self.active_event_loop = None
            liba8.monitor_summary()

    def get_current_state(self):
        liba8.get_current_state(self.output)

    def debugger_step(self):
        liba8.monitor_step()
        self.restart_cpu()

try:
    import wx

    class wxAtari800(Atari800):
        wx_to_akey = {
            wx.WXK_BACK: akey.AKEY_BACKSPACE,
            wx.WXK_DELETE: akey.AKEY_DELETE_CHAR,
            wx.WXK_INSERT: akey.AKEY_INSERT_CHAR,
            wx.WXK_ESCAPE: akey.AKEY_ESCAPE,
            wx.WXK_END: akey.AKEY_HELP,
            wx.WXK_HOME: akey.AKEY_CLEAR,
            wx.WXK_RETURN: akey.AKEY_RETURN,
            wx.WXK_SPACE: akey.AKEY_SPACE,
            wx.WXK_F7: akey.AKEY_BREAK,
            wx.WXK_PAUSE: akey.AKEY_BREAK,
            96: akey.AKEY_ATARI,  # back tick
        }

        wx_to_akey_ctrl = {
            wx.WXK_UP: akey.AKEY_UP,
            wx.WXK_DOWN: akey.AKEY_DOWN,
            wx.WXK_LEFT: akey.AKEY_LEFT,
            wx.WXK_RIGHT: akey.AKEY_RIGHT,
        }

        def process_key_down_event(self, evt):
            log.debug("key down! key=%s mod=%s" % (evt.GetKeyCode(), evt.GetModifiers()))
            key = evt.GetKeyCode()
            mod = evt.GetModifiers()
            if mod == wx.MOD_CONTROL:
                akey = self.wx_to_akey_ctrl.get(key, None)
            else:
                akey = self.wx_to_akey.get(key, None)

            if akey is None:
                evt.Skip()
            else:
                self.send_keycode(akey)

        def process_key_state(self):
            up = 0b0001 if wx.GetKeyState(wx.WXK_UP) else 0
            down = 0b0010 if wx.GetKeyState(wx.WXK_DOWN) else 0
            left = 0b0100 if wx.GetKeyState(wx.WXK_LEFT) else 0
            right = 0b1000 if wx.GetKeyState(wx.WXK_RIGHT) else 0
            self.input['joy0'] = up | down | left | right
            trig = 1 if wx.GetKeyState(wx.WXK_CONTROL) else 0
            self.input['trig0'] = trig
            #print "joy", self.emulator.input['joy0'], "trig", trig

            # console keys will reflect being pressed if at any time between frames
            # the key has been pressed
            self.input['option'] = 1 if wx.GetKeyState(wx.WXK_F2) else 0
            self.input['select'] = 1 if wx.GetKeyState(wx.WXK_F3) else 0
            self.input['start'] = 1 if wx.GetKeyState(wx.WXK_F4) else 0

except ImportError:
    class wxAtari800(object):
        def __init__(self, *args, **kwargs):
            raise RuntimeError("wx not available! Can't run wxAtari800")
